### ISSUE TITLE
Implement DataProcessors

### ISSUE THREAD

**Issue by gabizou** [SpongePowered/Sponge#8]
Given that Data API is rather large when it comes to the amount of data it is supposed to handle, this is a checklist of the data that is implemented.

If you'd like to implement one, please follow these guidelines by copy pasting this code block into your PR description and check mark as necessary:

----------

Implementation of `REPLACE_WITH_DATA_MANIPULATOR`.
- [x] Registration of field getters and setters
- [x] Accurately used the appropriate `AbstractData` implementation

Implementation of `REPLACE_WITH_IMMUTABLE_MANIPULATOR`
- [x] Accurately extended the appropriate `AbstractImmutableData` implementation
- [x] Accurately instantiating final instance fields with an `ImmutableValue` counter part for any value getters
- [x] If necessary, creating a new `ImmutableValue` for a value that should not be cached, or, using the existing caching utils.

Implementation of the `VALUE_PROCESSOR`(s):
- [x] Appropriately extended `AbstractSpongeValueProcessor`
- [x] Individual processors for each source type possible (one for `ItemStack`/`TileEntity`/`Entity` as necessary).

Registration:
- [x] Registered the `Key` correctly by `FIELD_NAME` in the `KeyRegistry`
- [x] Registered the `DataProcessor`s and `DataManipulatorBuilder` in `SpongeSerializationRegistry`
- [x] Registered the `ValueProcessor`s in the `SpongeSerializationRegistry`

You should create and include a test plugin in your pull request. Preferably, it should test all possible value related methods with your `DataManipulator`s and also test the serialization and deserialization of your `DataManipulator`s.

----------

The Checklist:

**TileData**
- [x] BannerData (@gabizou)
- [x] BeaconData (@Rain336)
- [x] BedData (@JBYoshi - replaced with `DyeableData`) - #1611
- [x] BrewingStandData (@Feeliiix)
- [x] CooldownData (@Rain336)
- [x] EndGatewayData (@kashike)
- [x] FurnaceData (@Feeliiix)
- [x] LockableData (@JBYoshi)
- [x] NoteData (@ZephireNZ)
- [x] SignData (@gabizou)
- [x] StructureData (@kashike)

**EntityData**
- [x] AbsorptionData (@kashike)
- [x] AchievementData (@ST-DDT / @windy1) - No longer relevant past 1.11
- [x] AffectsSpawningData (@kashike)
- [x] AgeableData (@Maxqia)
- [x] AgentData (@clienthax)
- [x] AggressiveData (@sanman00)
- [x] AngerableData (@sanman00)
- [x] AreaEffectCloudData (@gabizou)
- [x] ArmorStandData (@gabizou)
- [x] ArtData (@saladoc)
- [x] BodyPartRotationalData (@sanman00)
- [x] BreathingData
- [x] BreedableData (@oliverdunk)
- [x] CareerData
- [x] ChargedData (@clienthax)
- [x] CriticalHitData (@Rain336)
- [x] CustomNameVisibleData (@kashike)
- [x] DamageableData - Requires hooking into the CombatTracker (@Meronat)  - #1643
- [ ] DamagingData - Requires hooking into attacking.
- [x] DespawnDelayData (@kashike)
- [x] DominantHandData (@ZephireNZ) - #712
- [x] ElderData (@ZephireNZ) - No longer existing
- [x] ExperienceHolderData (@lucdon)
- [x] ExpirableData (@gabizou)
- [x] ExplosiveRadiusData (@windy)
- [x] ExpOrbData (@clienthax)
- [x] EyeLocationData (@DDoS)
- [x] FallDistanceData (@ZephireNZ)
- [x] FallingBlockData (@clienthax)
- [x] FlammableData (@JBYoshi)
- [x] FlyingAbilityData (@hsyyid)
- [x] FlyingData (@hsyyid)
- [x] FoodData (@gabizou)
- [x] FuseData (@windy1) 
- [x] GameModeData (@jamierocks)
- [x] GlowingData (@ZephireNZ)
- [x] GriefingData (@Cybermaxke)
- [x] HealingSourceData (@JBYoshi - replaced with `EnderDragon.getHealingCrystal()`)
- [x] HealthData
- [x] HealthScalingData (@gabizou)
- [x] HorseData (@Aaron1011)
- [x] IgniteableData (@hsyyid)
- [x] InvisibilityData (@hsyyid)
- [x] InvulnerabilityData (@JBYoshi)
- [x] JoinData (@gabizou)
- [x] KnockbackData (@sanman00)
- [ ] LeashData
- [x] MinecartBlockData (@ZephireNZ)
- [x] MovementSpeedData (@clienthax)
- [x] OcelotData (@sanman00)
- [x] ParrotData (@meronat)
- [x] PassengerData (@me4502)
- [x] PersistingData (@oliverdunk)
- [x] PigSaddleData (@clienthax)
- [x] PlayerCreatedData (@clienthax)
- [x] PlayingData (@clienthax)
- [x] RabbitData (@sanman00)
- [x] RespawnLocationData (@simon816)
- [x] ScreamingData (@Grinch)
- [x] ShatteringData (@clienthax)
- [x] ShearedData (@clienthax)
- [x] SittingData (@clienthax)
- [ ] SizeData (@hsyyid / @meronat) - https://github.com/SpongePowered/SpongeCommon/pull/1492
- [x] SkeletonData (@sanman00)
- [x] SkinData (@simon816)
- [ ] SleepingData (@sanman00) - https://github.com/SpongePowered/SpongeCommon/pull/1488
- [x] SlimeData (@Aaron1011)
- [x] SneakingData
- [x] SprintData (@kashike)
- [x] StatisticData (@ST-DDT / @windy1)
- [x] StuckArrowsData (@kashike)
- [x] TagData (@meronat) - https://github.com/SpongePowered/SpongeCommon/pull/1490
- [x] TameableData (@gabizou)
- [x] TargetLivingData (@Me4502)
- [x] TradeOfferData (@Rain336)
- [x] VehicleData (@hsyyid)
- [x] VelocityData (@gabizou)
- [x] VillagerZombieData (@Aaron1011)

**ItemData**
- [x] AuthorData (@lucdon)
- [x] BlockItemData (@kenzierocks)
- [x] BreakableData (@JBYoshi)
- [x] CoalData (@JBYoshi)
- [x] CookedFishData (@JBYoshi)
- [x] DurabilityData (@Rain336)
- [x] EnchantmentData (@gabizou)
- [x] FishData (@JBYoshi)
- [x] GenerationData (@JBYoshi)
- [x] GoldenAppleData (@Aaron1011)
- [x] HideData (@Rain336)
- [ ] InventoryItemData - Requires Inventory implementation
- [x] LoreData (@lucdon)
- [ ] MapItemData - Needs Map API in general
- [x] PagedData
- [x] PlaceableData (@JBYoshi)
- [x] SpawnableData (@Rain336)
- [x] SplashPotionData
- [x] StoredEnchantmentData (@Rain336)

**BlockData**
- [x] AttachedData (@lucdon)
- [x] AxisData (@lucdon)
- [x] BigMushroomData (@lucdon)
- [x] BrickData (@lucdon)
- [x] ComparatorData (@lucdon)
- [x] ConnectedDirectionData (@lucdon)
- [x] DecayableData (@lucdon)
- [x] DelayableData (@lucdon)
- [x] DirectionalData (@lucdon)
- [x] DirtData (@lucdon)
- [x] DisarmedData (@lucdon)
- [x] DisguisedBlockData (@lucdon)
- [x] DoublePlantData (@lucdon)
- [x] DropData (@lucdon)
- [x] ExtendedData (@lucdon)
- [x] FilledData (@Me4502)
- [x] FluidLevelData (@Me4502)
- [x] GrowthData (@lucdon)
- [x] HingeData (@lucdon)
- [x] InWallData (@lucdon)
- [x] LayeredData (@lucdon)
- [x] LogAxisData (@gabizou)
- [x] MoistureData (@Me4502)
- [x] OccupiedData (@lucdon)
- [x] OpenData (@lucdon)
- [x] PistonData (@lucdon)
- [x] PlantData (@lucdon)
- [x] PortionData (@lucdon)
- [x] PoweredData (@lucdon)
- [x] PrismarineData (@lucdon)
- [x] QuartzData (@lucdon)
- [x] RailDirectionData (@lucdon)
- [x] RedstonePoweredData (@lucdon)
- [x] SandData (@lucdon)
- [x] SandstoneData (@lucdon)
- [x] SeamlessData (@lucdon)
- [x] ShrubData (@lucdon)
- [x] SlabData (@lucdon)
- [x] SnowedData (@lucdon)
- [x] StairShapeData (@lucdon)
- [x] StoneData (@lucdon)
- [x] SuspendedData (@lucdon)
- [x] TreeData (@lucdon)
- [x] WallData (@lucdon)
- [x] WireAttachmentData (@JBYoshi) - https://github.com/SpongePowered/SpongeCommon/pull/1934

**Common Data**
- [x] ColoredData (@kenzierocks)
- [x] CommandData (@sanman00)
- [x] DisplayNameData
- [x] DyeableData (@Rain336)
- [x] FireworkEffectData (@ZephireNZ)
- [x] FireworkRocketData (@ZephireNZ)
- [x] MobSpawnerData (@ZephireNZ)
- [x] PotionEffectData (@gabizou)
- [x] RepresentedItemData
- [x] RepresentedPlayerData (@saladoc)
- [x] RotationalData (@Rain336)
- [x] SkullData
- [x] TargetedLocationData (@kashike)
- [x] WetData (@sanman00)

Updated as of January 30, 2018

**Comment by gabizou** [comment]
I am taking requests and offering the public to freely implement a DataProcessor and the associated ValueProcessors for the DataManipulator.

**Comment by RobertHerhold** [comment]
VelocityData please :)

**Comment by Aaron1011** [comment]
I'll take `GamemodeData`.

**Comment by mmonkey** [comment]
I second the VelocityData!

**Comment by gabizou** [comment]
I'm implementing `SignData` as we speak. I'll get to VelocityData likely tonight as well.

**Comment by ST-DDT** [comment]
I'll help with `StatisticData` and `AchievementData`.

**Comment by gabizou** [comment]
> I'll help with StatisticData and AchievementData.

Note that you'll have to implement far more than just the `StatisticDataProcessor` and `StatisticValueProcessor` since there is NO implementation of statistics in general in SpongeCommon.

**Comment by ST-DDT** [comment]
I know. I already mentioned it last weekend that I would start doing so before the data API merge.
I just checked in this issue, whether the data part is already implemented, and mentioned it here for documentation purposes.
However I cannot start immediately as I have to update my pending PRs first.

Thanks for your hint anyway.

**Comment by modwizcode** [comment]
I'll work on TradeOfferData.

**Comment by hsyyid** [comment]
I'd like to work on PassengerData

**Comment by DDoS** [comment]
Where's AABBData? It should be a common data (blocks and entities have it).

**Comment by hsyyid** [comment]
I'd like to work on `Whitelist Data`

**Comment by kamcio96** [comment]
I will work on TargetedLocationData. 
But DataManipulatorBuilder has create method. I don't know what is should return.

**Comment by ryantheleach** [comment]
Currently working on TameableData https://github.com/SpongePowered/SpongeCommon/pull/134

**Comment by gabizou** [comment]
Tagging @SpongePowered/SpongeAPI#812

**Comment by Aaron1011** [comment]
I'm taking `PotionEffectData`

**Comment by hsyyid** [comment]
I'm taking `IgniteableData`

**Comment by hsyyid** [comment]
I'd like to work on `FlyingData`

**Comment by lucdon** [comment]
i will work on LoreData

**Comment by JBYoshi** [comment]
I'll work on `BreakableData` and `PlaceableData`. https://github.com/JBYoshi/SpongeCommon/tree/data/placeable-breakable

**Comment by gabizou** [comment]
Just to alert people of various PR's and intentions to implement, there have been recent [changes](https://github.com/SpongePowered/SpongeCommon/commit/98504b97a1ef58ea1742e377bf86bc232becca75) to implementing `DataProcessor`s for `Entity` related data, so please, consider reading the changes.

**Comment by hsyyid** [comment]
Okay

**Comment by sanman00** [comment]
I'm working on ```WetData```.

**Comment by KAISERslin** [comment]
I really need "RepresentedItemData" to finish two of my open RPG Plugins.

**Comment by JBYoshi** [comment]
@KAISERslin `RepresentedItemData` was implemented with the recent `Block` data changes. af86add69d6cc75b618196461f2310ca9576b6b7

@gabizou Could you update this on the list?

**Comment by ZephireNZ** [comment]
`ElderData` implemented in #173

**Comment by gabizou** [comment]
After giving some thought with `DataProcessor`s and `ValueProcessor`s, I've decided to make a requirement for any and all PR's:

The PR must include a demo plugin (can be as simple as a listener listening for block changes that set the item in hand with various data) that uses all aspects of the processor implementations as expected (notwithstanding making API calls to try and retrieve data from known unsupported holders).

The reasoning behind this is that I've spent many many hours trying to make implementation easier to understand, but what I've found is that there's various mistakes made even when the author confirmed something was tested.

So, the PR itself will be reviewed with the following basis:

- The `DataManipulator` is implemented correctly using the appropriate abstraction of the sponge `DataManipulator` implementations.
- The `ImmutableDataManipulator` likewise is implemented correctly using the appropriate abstraction of the sponge `ImmutableDataManipulator` implementations.
- The `Key`s used by both manipulators are correctly registered in the `KeyRegistry`.
- The `DataManipulatorBuilder` for the `DataManipulator` is implemented and registered.
- The `DataProcessor` is implemented and uses the appropriate abstraction of `DataProcessor` for its uses.
- The `ValueProcessor` is implemented.
- None of the processors make method calls that would otherwise result in performing more lookups for the same processor that already existed (Examples include using `container.supports(key)` versus `supports(container)`).
- The demo plugin is easily copied and pasted into an IDE for live debugging.

Optionally:
- If the `DataManipulator` or data for the `Key`s being implemented are possible to exist in a `BlockState`, locating the proper [mixin](https://github.com/SpongePowered/SpongeCommon/blob/master/src/main/java/org/spongepowered/common/mixin/core/block/MixinBlockFlower.java#L46) to manually implement the data such that `BlockState`s support the newly implemented `Key`s and `DataManipulator`s is a plus
- If the data being implemented is also supported by any `Entity` in the entity hierarchy (like a `Sheep` supporting `DyeableData`), locating the proper mixin to not only [override the necessary method](https://github.com/SpongePowered/SpongeCommon/blob/master/src/main/java/org/spongepowered/common/interfaces/entity/IMixinEntity.java#L57) to provide said data by default when an `Entity#toContainer()` is called is a plus.

Overall, my goal has been to try and help people understand the various systems and structures set in place to not only implement getting/setting data through the API, but have it work in all of it's designed intentions. Unfortunately, the only way I can see some people learning how and why certain things are implemented the way they are is by putting these types of requirements.

I hope that this will provide a learning experience, not to make reviewing PR's easier, but to help teach everyone what is required to implement certain aspects of the API to help future contributions require less code cleanup and less sanity testing.

The following PR's are ones I've selected that need to meet these guidelines:
- #135 
- #136 
- #139 
- #143
- #144 
- #148 
- #159 
- #165 
- #166 
- #173

Tomorrow, I will comment with the checklist on each of these PR's. If a PR is significantly old enough that the author is taking too long, I will manually fix the minor changes required and merge. However, if a PR is still awaiting for the author to respond after a while, the PR may be disregarded and re-implemented by a member of the dev-team.

**Comment by lucdon** [comment]
I'm working on ExperienceHolderData

**Comment by ST-DDT** [comment]
I have a question regarding the Data API.

How can mods/plugins make their custom entities return existing data instances for data that are not supported over inheritance.

Example. A new special protester/demonstrator villager carries a (readable) sign with some text on it.
The plugin author wants to use SignData for this for obvious reasons. However the registered DataProcessors only support blocks. How does he support this, without rewriting the entire data implementation for this entity.

**Comment by gabizou** [comment]
> How can mods/plugins make their custom entities return existing data instances for data that are not supported over inheritance.

Multiple processor registration. If you have your own custom entity, you're already interacting with forge, and what you can do is register the processors directly in your mod or in a compatibility module. SpongeCommon already handles multiple processor registration quite well (it's how it works already).

**Comment by ST-DDT** [comment]
@gabizou Thanks.
One more thing, could you check/flag the recently merged data classes? (Ignite, Sneaking...)

**Comment by JBYoshi** [comment]
@gabizou `EnchantmentData`, `GoldenAppleData`, `IgniteableData`, `LoreData`, `PagedData`, `RepresentedItemData`, and `SneakingData` have been implemented. Could you update these on the list?

**Comment by Tzky** [comment]
``FlyingData`` was just pushed too.

**Comment by JBYoshi** [comment]
Implementing `CoalData`, `CookedFishData`, and `FishData`, please wait...

**Comment by gabizou** [comment]
I'm going to make some minor changes to the implementation regarding getting cached values. The functionalities will still remain the same, nothing will break code wise, but getting cached values will be easier.

Namely, what will be doable is the following:
```java
final ImmutableValue<Boolean> canFly = ImmutableSpongeValue.cachedOf(Keys.CAN_FLY, false, actualValue);
```

**Comment by lucdon** [comment]
implemented AuthorData while listening to SoS6

**Comment by hsyyid** [comment]
I'll do the value processor for Keys.CAN_FLY

**Comment by Aaron1011** [comment]
I'm taking `FuseData`

**Comment by hsyyid** [comment]
I'm taking `InvisibilityData`

**Comment by Feeliiix** [comment]
Started work in the FurnaceData

**Comment by gabizou** [comment]
I've updated the issue description for what I'd like to see in every Data PR as the top PR description (the first post).

**Comment by clienthax** [comment]
Taking SittingData

**Comment by Aaron1011** [comment]
I'm taking VillagerZombieData.

**Comment by Feeliiix** [comment]
I'm taking the BrewingData

**Comment by hsyyid** [comment]
I'm taking `SizeData`

**Comment by gabizou** [comment]
Updated the list as of this comment.

**Comment by gabizou** [comment]
> I will work on TargetedLocationData. 
> But DataManipulatorBuilder has create method. I don't know what it should return.

Any updates on this @kamcio96? It's been two months and haven't seen any PR's for `TargetedLocationData`.

**Comment by lucdon** [comment]
will implement stone and prismarine data #206

**Comment by hsyyid** [comment]
I'm taking `MapItemData`

**Comment by gabizou** [comment]
> I'm taking MapItemData

Note that you'll have to write and implement Map API as well for that to work.

**Comment by hsyyid** [comment]
@gabizou Yep

**Comment by kenzierocks** [comment]
I will dutifully take `ColoredData`.

**Comment by SocraticPhoenix** [comment]
Rather than attempt to implement one and fail horribly, I'll politely request that AgentData be implemented to the point that we can enable/disable AI.

**Comment by clienthax** [comment]
Taking AgentData

**Comment by gabizou** [comment]
List is updated to here.

**Comment by me4502** [comment]
I'm claiming the following:

MoistureData
FluidLevelData
FilledData

**Comment by clienthax** [comment]
Taking OpenData

**Comment by lucdon** [comment]
@me4502 im trying to finish my pr today and else tomorrow

**Comment by sanman00** [comment]
I'm taking `LockableData`.

**Comment by clienthax** [comment]
taking ChargedData

**Comment by clienthax** [comment]
taking PlayerCreatedData

**Comment by hsyyid** [comment]
Taking `OccupiedData`

**Comment by clienthax** [comment]
Taking ShatteringData

**Comment by hsyyid** [comment]
`InWallData` was also done..

**Comment by lucdon** [comment]
working on attributedata and the attribute implementation(and some api changes)

**Comment by ZephireNZ** [comment]
@lucdon Attributes got removed from the API in SpongePowered/SpongeAPI@a6391b6

**Comment by lucdon** [comment]
@ZephireNZ ahh thx did not update my api fork so i didnt see it.
@gabizou can you remove attribute data from this issue

**Comment by me4502** [comment]
I'm going to hate myself for this, but I'll take FireworkData as well.

**Comment by Feeliiix** [comment]
FurnaceData and BrewingStandData finished!

Taking FlamableData now!

**Comment by lucdon** [comment]
AttachedData
 AxisData
 ConnectedDirectionData
 DirectionalData
 DisarmedData
 DropData
 ExtendedData
 GrowthData
 OpenData
 PoweredData
 RedstonePoweredData
 SeamlessData
 SignaledOutputData
 SnowedData
 SuspendedData

taking the 15 remaining block data manipulators

**Comment by ZephireNZ** [comment]
@me4502 have you made progress on fireworks? I got partially into implementing it before I saw you your comment haha

**Comment by me4502** [comment]
@ZephireNZ I haven't started it yet, I've been waiting on my first data PR to be accepted before doing more.

**Comment by ZephireNZ** [comment]
Did you want to keep it, or shall I do it? :)

**Comment by me4502** [comment]
@ZephireNZ You can do it, I only grabbed it as I wanted it done, and no one else had grabbed it.

**Comment by ZephireNZ** [comment]
Sweet :)

**Comment by kenzierocks** [comment]
I'll take `BlockItemData`.

**Comment by gabizou** [comment]
Updated the description up to here.

**Comment by gabizou** [comment]
Also, a correction: `OccupiedData` may be implemented in that the manipulator exists, but the actual implementation backing it is not done. So, it's still up for grabs, maybe @lucdon ?

**Comment by clienthax** [comment]
@SocraticPhoenix AgentData is implemented and merged

**Comment by gabizou** [comment]
@clienthax, yes, as the description said, it was implemented...

**Comment by JBYoshi** [comment]
@gabizou What's the difference between `AngerableData` and `AggressiveData`?

**Comment by ZephireNZ** [comment]
@JBYoshi Angerable is for mobs like pigmen I believe. As for Aggressive I believe that's for "hostile" mobs

**Comment by SocraticPhoenix** [comment]
@clienthax thanks
Now all I need is InvisibilityData to be completed and my NPCs will be complete!

**Comment by hsyyid** [comment]
@SocraticPhoenix I'm working on it :wink:

**Comment by lucdon** [comment]
will implement 'OccupiedData' with the rest of the blockData

**Comment by lucdon** [comment]
will implement these 3 too because only the manipulators are created and nothing more(and @hsyyid  thought the were implemented which is false)
InWallData
LayeredData
DecayableData

**Comment by gabizou** [comment]
So, just to make sure, @hsyyid isn't going to be implement those?

**Comment by JBYoshi** [comment]
Taking `LockableData`.

**Comment by SocraticPhoenix** [comment]
I'd like to request that FallingBlockData be implemented

**Comment by gabizou** [comment]
Updated to here. Added SplashPotionData.

**Comment by saladoc** [comment]
`SignaledOutputData` is on the list, but does not seem to exist in the API.

**Comment by hsyyid** [comment]
@gabizou No. @lucdon is correct, I thought they were done, but they're only half done. I am fine with @lucdon doing them instead, as he's already done most if not all blcokdata processors.

**Comment by lucdon** [comment]
@gabizou what about creating delayableData for RedstoneRepeaters?

**Comment by gabizou** [comment]
Yep, go ahead and add it @lucdon

**Comment by clienthax** [comment]
Taking FallingBlockData

**Comment by SocraticPhoenix** [comment]
@clienthax Thank you!

**Comment by sanman00** [comment]
I'm working on `SkeletonData`.

**Comment by Feeliiix** [comment]
Hey guyz,

i started to implement the <code>FlammableData</code>, but in my opinion it does not store what it has to store. At the moment <code>FlammableData</code> returns is a entity is aflame, but i think flammable means if something could burn and not if something is aflame.

There are 2 reasons i think:
1. Rename <code>FlammableData</code> to something like <code>AFlameData</code>
2. Create a 2nd data named like "AFlameData" and let <code>FlammableData</code> store was it has to store.

Thanks for your opinion @gabizou @Aaron1101 and all other guyz.

**Comment by JBYoshi** [comment]
@Feeliiix Any entity that [supports](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/data/value/mutable/CompositeValueStore.java#L149) `FlammableData` can be set aflame.

**Comment by Feeliiix** [comment]
@JBYoshi right but in my opinion than it should be store a boolean if it support that and not a boolean that represents the aflame state
<code>
/**
     * Gets the {@link Value} for the "aflame" state.
     *
     * @return The value for the aflame state
     */
</code>

**Comment by ST-DDT** [comment]
There was a discussion for this on IRC lately, but it did not have an answer.
The suggestion there was:
`BurningData` for things that are on fire and `FlammableProperty` for things that could be on fire.

**Comment by gabizou** [comment]
The way I thought of it is that since it only stores the "ticks" of being on fire, it's only available when the entity is actually on fire. If the entity is not on fire, then there's no flamabledata to get. Also, I'd appreciate not cluttering up this issue in the future to take it to a new issue, possibly in API as it most directly affects that more than anything.

**Comment by Feeliiix** [comment]
@gabizou I See your point, but i suggesting a rename because the name is very cunfusingly, when it does only store the ticks, the entity is aflame

**Comment by clienthax** [comment]
@SocraticPhoenix FallingBlockData is in master now

**Comment by SocraticPhoenix** [comment]
@clienthax 
thank you

**Comment by sanman00** [comment]
I'm working on `OcelotData` and `RabbitData` alongside `SkeletonData`.

**Comment by gabizou** [comment]
Updated the checklist as of now.

**Comment by ZephireNZ** [comment]
I'm taking MobSpawnerData

**Comment by gabizou** [comment]
@ZephireNZ you'll have to wait for @Deamon's world gen PR

**Comment by ZephireNZ** [comment]
Can do, which part of his PR am I needing?

**Comment by Deamon5550** [comment]
@ZephireNZ The WeightedTable and related value changes in api and common.

**Comment by sanman00** [comment]
I'm working on `AggressiveData`.

**Comment by ryantheleach** [comment]
Keen to see AgeableData implemented soon, if someone tackles it, could they also consider keys for BABY / ADULT?

**Comment by sanman00** [comment]
I'm working on `SplashPotionData`.

**Comment by JBYoshi** [comment]
I'll write `CloneableData` and upload a *copy* of it. (Pun intended.)

**Comment by gabizou** [comment]
@JBYoshi, it's now GenerationData

**Comment by JBYoshi** [comment]
Okay, thanks.

**Comment by Aaron1011** [comment]
Implementing `remove` properly can sometimes be tricky to get right. When writing it, keep these things in mind:

(Note: there is an [open issue](https://github.com/SpongePowered/SpongeCommon/issues/332) regarding the behavior of `DisplayNameData` with relation to `Human`s. For the purpose

---------

1. Does the data in question always exist (even if it's some 'empty' or 'null' value)? If so, `remove` should **not** set it to this value. Instead, it should return `DataTransactionResult.failNoData`.

    Take `DisplayNameData`, for example. On a `Humanoid` (a `Player` or `Human`), this data will  always exist. At best, you can set the display name to the empty string (or `Texts.of()`), but it still exists! You can see this for yourself by setting the display name of a Human to `Texts.of()`, and observing that a nameplate still appears over its head in game.

   In the case of an `ItemStack`, however, the data does *not* always exists. If the `ItemStack` does not have an NBT compound, or does not have the proper key set in it, the data is not present at all. Thus, removing it makes sense, since there's no 'empty' or 'null' value that gets set.

2. Does calling `remove` immediately followed by `get` return a present `Optional` (if removing the data is supported)? If so, this is definitely a bug, and likely a sign that removing the data only sets it to some default value.

3. Can the behavior of `remove` be replicated through normal usage of the Data API by a plugin? This is similar to the above case, and means that `remove` isn't actually removing anything. It should only be supported in the case where a meaningful way exists to 'delete' the data from `DataHolder`, such that it no longer exists at all.

    In the `DisplayNameData` example, this would mean that a plugin could remove the data of a (non-`Human`) `Entity` by setting its display name to the empty `Text`. This won't work, since a non-present display name is different from one set to the empty `Text`.

    In the case of a `Human`,  this is a non-issue, since `DisplayNameData` cannot be removed at all, by `remove` or otherwise.

---------

Hopefully, these questions will help to clear up any confusion as to whether `remove` should be implemented for a given `DataProcessor`.

**Comment by gabizou** [comment]
@saladoc @Tzky probably want to include the above in the contributing docs.

**Comment by sanman00** [comment]
I'm working on `CommandData`.

**Comment by gabizou** [comment]
Updated the checklist as of now. @ZephireNZ the mob spawner changes are updated, including the implementation of the data manipulator itself.

**Comment by sanman00** [comment]
I'm working on `KnockbackData`.

**Comment by JBYoshi** [comment]
@Rain336 is working on `DyeableData` (#359).

**Comment by oliverdunk** [comment]
Tick off PersistantData? (#370)

**Comment by kashike** [comment]
`AgeableData` is mine

**Comment by sanman00** [comment]
I'm working on `BodyPartRotationalData`.

**Comment by kashike** [comment]
`SprintData` is mine: #378

**Comment by oliverdunk** [comment]
I'm working on BreedableData

**Comment by JBYoshi** [comment]
Taking most of the remaining entity data:
- `AgeableData`
- `CriticalHitData`
- `ExplosiveRadiusData`
- `FlammableData`
- `GriefingData`
- `HealingSourceData`
- `InvulnerabilityData`
- `PersistingData`

@gabizou Could you remove `LeashData` from the list since it's been replaced with `Entity` methods and `CloneableData` since it's been renamed (and the new name, `GenerationData`, is already on the list)?

**Comment by kashike** [comment]
@JBYoshi `PersistingData` is taken already

**Comment by JBYoshi** [comment]
@kashike Okay, removing that from my todo list.

**Comment by JBYoshi** [comment]
So is `CriticalHitData`.

**Comment by saladoc** [comment]
@JBYoshi I'm currently giving `ExplosiveRadiusData` a try, but did not post here. Could you move it to the bottom of your list maybe? Then you can still take it if my efforts fail

**Comment by JBYoshi** [comment]
Sure.

**Comment by kashike** [comment]
I'm reclaiming unimplemented potion data.

**Comment by Rain336** [comment]
I'm working on `WireAttachmentData`

**Comment by sanman00** [comment]
I'll be working on `FuseData` since it does not appear to be implemented and I haven't found a pull request for it.
@Aaron1011 is this fine with you?

**Comment by Rain336** [comment]
I implemented HideData(#388)

**Comment by Aaron1011** [comment]
@sanman00: Sure! :)

**Comment by saladoc** [comment]
I unclaim `ExplosiveRadiusData` because I feel like it would keep me from the docs too long. See SpongePowered/SpongeAPI#1008 for my findings/problems.

**Comment by gabizou** [comment]
Description/checklist updated as of this comment.

**Comment by hsyyid** [comment]
@JBYoshi Can I take `AgeableData`off of your plate?

**Comment by JBYoshi** [comment]
@hsyyid Sure.

**Comment by gabizou** [comment]
I've freed up `TargetedLocationData` and `BeaconData`.

**Comment by clienthax** [comment]
taking BeaconData

**Comment by JBYoshi** [comment]
@clienthax In your `BeaconData` PR, could you also update [line 51 in `SpongeBeaconBuilder`](https://github.com/SpongePowered/SpongeCommon/blob/master/src/main/java/org/spongepowered/common/data/builder/block/tileentity/SpongeBeaconBuilder.java#L51)?

**Comment by clienthax** [comment]
@JBYoshi yep

**Comment by JBYoshi** [comment]
@clienthax Thanks. Eclipse (with my custom compiler error/warning settings) has been nagging me about a few null errors (due to unimplemented data) for the past few months, and this is the last of them.

**Comment by clienthax** [comment]
Yeah screw beacondata, stupid freking optional bs.

**Comment by ryantheleach** [comment]
I'd like to request JoinData it's currently blocking any form of setting the spawn for players logging on for the first time, as it's currently not firing RespawnEvents(they were never dead, just not born)

**Comment by randombyte-developer** [comment]
I wanted to use TargetedLocationData, read the list of this issue and saw that TargetedLocationData isn't done. So I waited and waited, nothing happened. Today I wanted to implement it myself and saw [this PR](https://github.com/SpongePowered/SpongeCommon/pull/485).
TargetedLocationData should be ticked in the list above.

**Comment by kashike** [comment]
@gabizou This list needs updating whenever you get time :)

**Comment by simon816** [comment]
I have updated the list

**Comment by ryantheleach** [comment]
Cheers simon! It's damned close to being closed.

**Comment by kenzierocks** [comment]
I'm going to take WireAttachmentData.

**Comment by gravityfox** [comment]
DisplayNameData does not work for Itemstacks. It doesn't seem to be fully implemented.

https://github.com/SpongePowered/SpongeCommon/blob/master/src/main/java/org/spongepowered/common/data/processor/data/DisplayNameDataProcessor.java#L129

The key `Keys.DISPLAY_NAME` seems to work fine though, for some reason.

**Comment by ZephireNZ** [comment]
@gabizou Beds don't have a manipulator for their "part" property.

They're different enums, but should we just make it use `PortionData` instead? We could add a note that "TOP" corresponds to "HEAD" etc.

**Comment by kashike** [comment]
This has been updated, thanks to @Meronat.

**Comment by parlough** [comment]
I will start BedData and possibly LeashData later today or tomorrow.

**Comment by ghost** [comment]
I am working on `DominantHandData`.

**Comment by ryantheleach** [comment]
@Meronat @sanman00 How are the data's going? Not many left!

**Comment by parlough** [comment]
Working on it as time allows. Have a few PRs open implementing various data that should be good to go. Will be opening a PR to add disabled slot control to ArmorStandData soon, then I'll look into bed and leashes. Then from there we will see, I may look into the missing combat ones.

**Comment by JBYoshi** [comment]
What's left for `WireAttachmentData`?

**Comment by ryantheleach** [comment]
The ShortList:

- [ ] DamagingData - Requires hooking into attacking.
- [x] DominantHandData (@ZephireNZ / @Meronat) - #712
- [ ] LeashData
- [ ] SizeData (@hsyyid / @Meronat) - #1492
- [ ] SleepingData (@sanman00) - #1488
- [x] TagData (@Meronat) - #1490
- [ ] InventoryItemData - Requires Inventory implementation
- [ ] MapItemData - Needs Maps API in general
- [x] WireAttachmentData - Partially done

**Comment by gabizou** [comment]
@ryantheleach Maps API was abandoned ~again~, so it's something that modwiz isn't going to be getting to anytime soon.

**Comment by XakepSDK** [comment]
`Keys.DISPLAY_NAME` does not work for players.

**Comment by kashike** [comment]
...yes it does? If it didn't we'd have 100 issues opened by now.

**Comment by JBYoshi** [comment]
I'm finishing up `WireAttachmentData`. https://github.com/SpongePowered/SpongeCommon/pull/1934

**Comment by Zidane** [comment]
Closing as this is more or less complete.
